### PR TITLE
ensure the XContentBuilder is always closed in RestBuilderListener

### PR DESCRIPTION
**Pull request by jaymode**
There may be cases where the XContentBuilder is not used and therefore it never gets closed, which can cause a leak of bytes. This change moves the creation of the builder into a try with resources block.

### CODE REVIEW

**Review by dadoonet — APPROVED** [review]

**Comment by dadoonet** [comment]
I think this should go as well in 5.0 branch, no?

**Comment by bleskes** [comment]
I'm not sure we want to automatically close. The current api is set up so that we expect to use the builder we give them (there is no way to  not have a builder) - if they don't it's probably a bug we're hiding (or we should give an official way to not have a builder created at all). I think we should add an assertion that verifies the builder is closed when buildResponse returns

**Comment by s1monw** [comment]
@bleskes maybe both, assertion and auto close? I mean some plugins might not test this (3rd party ) and then we are leaking bytes since assertions are off?

**Comment by bleskes** [comment]
@s1monw yeah , that's what I meant ++ . Keep code as is but make sure builder is closed when buildResponse returns

**Comment by jaymode** [comment]
> I think this should go as well in 5.0 branch, no?

I think that makes sense, I added the 5.0.1 label.

> maybe both, assertion and auto close?

I pushed [5a17d14](https://github.com/elastic/elasticsearch/pull/21124/commits/5a17d141e1cffffd328387dfddef6b17c1628963), which adds the assertion while keeping auto close.

File: core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java [hunk 1]
```diff
@@ -28,19 +28,25 @@
...
         super(channel);
     }
 
     @Override
     public final RestResponse buildResponse(Response response) throws Exception {
-        try (XContentBuilder builder = channel.newBuilder()){
-            return buildResponse(response, builder);
+        try (XContentBuilder builder = channel.newBuilder()) {
+            final RestResponse restResponse = buildResponse(response, builder);
+            assert assertionsEnabled == false || builder.isClosed() : "callers should ensure the XContentBuilder is closed themselves";
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java]: can you add a protected method `assert assertBuilderClosed(builder)` we can override in tests instead. this is very confusing and I hate static mutable vars

File: core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java [hunk 2]
```diff
@@ -998,6 +998,10 @@ public void close() {
         }
     }
 
+    public boolean isClosed() {
+        return generator.isClosed();
+    }
+
     public XContentGenerator generator() {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/xcontent/XContentBuilder.java]: would it be enough to only add `isClosed` to XContentGenerator? and maybe add a javadoc?

**Review by s1monw — CHANGES_REQUESTED** [review]
left two comments

File: core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java (hunk 1 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java]: ++ that's a much better approach

File: core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java [hunk 3]
```diff
@@ -49,4 +46,9 @@ public final RestResponse buildResponse(Response response) throws Exception {
      * using the {@link XContentBuilder#close()} method.
      */
     public abstract RestResponse buildResponse(Response response, XContentBuilder builder) throws Exception;
+
+    // pkg private method that we can override for testing
+    boolean assertBuilderClosed(XContentBuilder xContentBuilder) {
+        return xContentBuilder.generator().isClosed();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/action/RestBuilderListener.java]: I usually do this:
```
assert  xContentBuilder.generator().isClosed();
return true;
```

**Review by s1monw — APPROVED** [review]

**Comment by s1monw** [comment]
left a minor LGTM otherwise

**Comment by s1monw** [comment]
@jaymode can we get this in please?

**Pull request merged**